patroni/postgresql/sync.py

```diff
@@ -239,52 +260,127 @@ def current_state(self, cluster: Cluster) -> Tuple[CaseInsensitiveSet, CaseInsen
-                    sync_nodes.add(app_name)
-            if len(candidates) >= sync_node_count:
-                break
-
-        return candidates, sync_nodes
-
-    def set_synchronous_standby_names(self, sync: Collection[str]) -> None:
+                if self._postgresql.global_config.is_quorum_commit_mode:
+                    # add nodes with nofailover tag only to get enough "active" nodes
+                    if not nofailover or len(active) < sync_node_count:
+                        if app_name in self._ready_replicas:
+                            numsync_confirmed += 1
+                        active.add(app_name)
+                else:
+                    active.add(app_name)
+                    if sync_state == 'sync' and app_name in self._ready_replicas:
+                        sync_nodes.add(app_name)
+                        numsync_confirmed += 1
+                    if len(active) >= sync_node_count:
+                        break
+
+        if self._postgresql.global_config.is_quorum_commit_mode:
+            sync_nodes = CaseInsensitiveSet() if self._ssn_data.has_star else self._ssn_data.members
+
+        return _SyncState(
+            self._ssn_data.sync_type,
+            0 if self._ssn_data.has_star else self._ssn_data.num,
+            numsync_confirmed,
+            sync_nodes,
+            active)
+
+    def set_synchronous_standby_names(self, sync: Collection[str], num: Optional[int] = None) -> None:
         """Constructs and sets "synchronous_standby_names" GUC value.
 
         :param sync: set of nodes to sync to
+        :param num: specifies number of nodes to sync to. The *num* is set only in case if quorum commit is enabled
         """
-        has_asterisk = '*' in sync
+        # Special case. If sync nodes set is empty but requested num of sync nodes >= 1
+        # we want to set synchronous_standby_names to '*'
+        has_asterisk = '*' in sync or num and num >= 1 and not sync
         if has_asterisk:
             sync = ['*']
         else:
-            sync = [quote_ident(x) for x in sync]
+            sync = [quote_ident(x) for x in sorted(sync)]
 
         if self._postgresql.supports_multiple_sync and len(sync) > 1:
-            sync_param = '{0} ({1})'.format(len(sync), ','.join(sync))
+            if num is None:
+                num = len(sync)
+            sync_param = ','.join(sync)
         else:
             sync_param = next(iter(sync), None)
 
+        if TYPE_CHECKING:  # pragma: no cover
+            assert self._postgresql.global_config is not None
+
+        if self._postgresql.global_config.is_quorum_commit_mode and sync or\
+                self._postgresql.supports_multiple_sync and len(sync) > 1:
+            prefix = 'ANY ' if self._postgresql.global_config.is_quorum_commit_mode\
+                and self._postgresql.supports_quorum_commit else ''
+            sync_param = '{0}{1} ({2})'.format(prefix, num, sync_param)
```

Comment:
Yes, agreed. Perhaps it reads better when function call results are stored in variables or variables are used more than once?

I agree, it's not a hard rule, there would be situations where it would be better to use another form. For instance, it is generally seen as better to use `logging`'s own string formatting: `logging.info("This is my %s", message)`.